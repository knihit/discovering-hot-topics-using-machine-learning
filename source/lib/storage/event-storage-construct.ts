#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { IDatabase } from "@aws-cdk/aws-glue";
import { Effect, Policy, PolicyStatement, Role, ServicePrincipal } from "@aws-cdk/aws-iam";
import { CfnDeliveryStream, CfnDeliveryStreamProps } from "@aws-cdk/aws-kinesisfirehose";
import { Code, Function, Runtime } from "@aws-cdk/aws-lambda";
import { Bucket } from '@aws-cdk/aws-s3';
import { Aws, Construct } from "@aws-cdk/core";
import { KinesisFirehoseToS3 } from '@aws-solutions-constructs/aws-kinesisfirehose-s3';
import { buildLambdaFunction } from '@aws-solutions-constructs/core';

export interface EventStorageProps {
    readonly compressionFormat: string,
    readonly prefix?: string,
    readonly processor?: boolean,
    readonly s3Bucket: Bucket,
    readonly keyArn?: string,

    //below props only required if data format is parquet
    /**
     * Convert data to Apache Parquet format
     */
     readonly convertData?: boolean,

    /**
     * The database name is required if Firehose should convert data to Apache Parquet
     */
    readonly database?: IDatabase,

    /**
     * Table is required to add dependency for Firehose creation
     */
    readonly tableName?: string
}

export class EventStorage extends Construct {

    private readonly _lambda: Function;
    private _firehose: CfnDeliveryStream;

    constructor (scope: Construct, id: string, props: EventStorageProps) {
        super(scope, id);

        // Setup the IAM Role for Kinesis Firehose
        const firehoseRole = new Role(this, 'FirehoseRole', {
            assumedBy: new ServicePrincipal('firehose.amazonaws.com'),
        });

        props.s3Bucket.grantReadWrite(firehoseRole); // add permissions to read/write to S3 bucket

        let firehoseGluePolicy: Policy;
        if (props.convertData && props.tableName !== undefined) {
            firehoseGluePolicy = new Policy(this, 'FirehoseGlueTablePolicy', {
                statements: [ new PolicyStatement({
                    effect: Effect.ALLOW,
                    actions: [ 'glue:GetTable', 'glue:GetTableVersion', 'glue:GetTableVersions' ],
                    resources: [
                        `arn:aws:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:table/${props.database!.databaseName}/${props.tableName}`,
                        `arn:aws:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:database/${props.database!.databaseName}`,
                        `arn:aws:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:catalog`
                    ]
                })]
            })

            firehoseGluePolicy.attachToRole(firehoseRole);
        }

        let firehoseGlueKmsPolicy: Policy;
        if (props.keyArn !== undefined) {
            firehoseGlueKmsPolicy = new Policy(this, 'FirehoseGlueKms', {
                    statements: [ new PolicyStatement({
                    effect: Effect.ALLOW,
                    resources: [ props.keyArn ],
                    actions: [ 'kms:Decrypt' ]
                })]
            });

            firehoseGlueKmsPolicy.attachToRole(firehoseRole);
        }

        if (props?.processor) {
            this._lambda = buildLambdaFunction(this, {
                lambdaFunctionProps: {
                    runtime: Runtime.NODEJS_12_X,
                    handler: 'index.handler',
                    code: Code.fromAsset(`${__dirname}/../../lambda/storage-firehose-processor`),
                }
            });

            const lambdaProcessorPolicy = new Policy(this, 'LambdaProcessor', {
                statements: [new PolicyStatement({
                    effect: Effect.ALLOW,
                    resources: [ this._lambda.functionArn ],
                    actions: [ "lambda:InvokeFunction", "lambda:GetFunctionConfiguration"  ]
                })]
            });

            lambdaProcessorPolicy.attachToRole(firehoseRole);
        }

        // Setup the default Kinesis Firehose props
        const defaultKinesisFirehoseProps: CfnDeliveryStreamProps = {
            extendedS3DestinationConfiguration : {
                bucketArn: props.s3Bucket.bucketArn,
                bufferingHints: {
                    intervalInSeconds: 600,
                    sizeInMBs: 64
                },
                compressionFormat: props.compressionFormat,
                roleArn: firehoseRole.roleArn,
                ...(props.prefix && {
                    prefix: `${props?.prefix}created_at=!{timestamp:yyyy-MM-dd}/`,
                    errorOutputPrefix: `${props.prefix}error/!{firehose:random-string}/!{firehose:error-output-type}/created_at=!{timestamp:yyyy-MM-dd}/`
                }),
                ...(props?.processor && {
                        processingConfiguration: {
                        enabled: true,
                        processors: [{
                            type: 'Lambda',
                            parameters: [{
                                parameterName: 'LambdaArn',
                                parameterValue: this._lambda.functionArn
                            }]
                        }]
                    }
                }),
                ...(props.convertData && {
                    dataFormatConversionConfiguration: {
                        inputFormatConfiguration: {
                            deserializer: {
                                openXJsonSerDe: {}
                            }
                        },
                        outputFormatConfiguration: {
                            serializer: {
                                parquetSerDe: {}
                            }
                        },
                        schemaConfiguration: {
                            databaseName: props.database!.databaseName,
                            tableName: props.tableName,
                            roleArn: firehoseRole.roleArn
                        }
                    }
                })
            }
        }

        const firehoseToS3 = new KinesisFirehoseToS3(this, 'KinesisFirehose', {
            kinesisFirehoseProps: defaultKinesisFirehoseProps,
            existingBucketObj: props.s3Bucket
        });

        this._firehose = firehoseToS3.kinesisFirehose;
        this._firehose.node.addDependency(firehoseRole);
        if (props.convertData && props.tableName !== undefined) {
            this._firehose.node.addDependency(firehoseGluePolicy!);
            this._firehose.node.addDependency(firehoseGlueKmsPolicy!);
        }
    }

    public get kinesisFirehose(): CfnDeliveryStream {
        return this._firehose;
    }

    public get deliveryStreamName(): string {
        return this.kinesisFirehose.ref
    }

    public get deliveryStreamArn(): string {
        return this.kinesisFirehose.attrArn;
    }
}