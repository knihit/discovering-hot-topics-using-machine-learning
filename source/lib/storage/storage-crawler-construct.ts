#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Schedule } from "@aws-cdk/aws-events";
import { Effect, Policy, PolicyStatement } from "@aws-cdk/aws-iam";
import { Code, Runtime } from '@aws-cdk/aws-lambda';
import { Bucket } from "@aws-cdk/aws-s3";
import { Aws, Construct, CustomResource, Duration } from "@aws-cdk/core";
import { EventsRuleToLambda } from '@aws-solutions-constructs/aws-events-rule-lambda';
import { buildLambdaFunction } from "@aws-solutions-constructs/core";

export interface StorageCrawlerProps {
    readonly s3Bucket: Bucket,
    readonly databaseName: string,
    readonly keyArn: string,
    readonly tableMap: Map<string, string>
}

export class StorageCrawler extends Construct {
    constructor(scope: Construct, id: string, props: StorageCrawlerProps) {
        super(scope, id);

        const lambdaGluePolicy = new Policy(this, 'LambdaGlue', {
            statements: [
                new PolicyStatement({
                    effect: Effect.ALLOW,
                    resources: [
                        `arn:${Aws.PARTITION}:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:table/${props.databaseName}/*`,
                        `arn:${Aws.PARTITION}:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:database/${props.databaseName}`,
                        `arn:${Aws.PARTITION}:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:catalog`,
                    ],
                    actions: [
                        'glue:GetPartition', 'glue:GetTable', 'glue:CreatePartition'
                    ]
                }),
                new PolicyStatement({
                    effect: Effect.ALLOW,
                    resources: [ props.keyArn ],
                    actions: [ 'kms:Decrypt' ]
                }),
                new PolicyStatement({
                    effect: Effect.ALLOW,
                    actions: [ 'logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents' ],
                    resources: [ `arn:${Aws.PARTITION}:logs:${Aws.REGION}:${Aws.ACCOUNT_ID}:/aws-glue/*` ]
                })
            ]
        });

        const lambdaFunc = buildLambdaFunction(this, {
            lambdaFunctionProps: {
                description: 'Lambda function to create parition on glue tables',
                runtime: Runtime.NODEJS_12_X,
                handler: 'index.handler',
                code: Code.fromAsset(`${__dirname}/../../lambda/create-partition`),
                timeout: Duration.minutes(5),
                environment: {
                    DATABASE_NAME: props.databaseName,
                    TABLE_NAMES: Array.from(props.tableMap.values()).join(',')
                }
            }
        });

        const createParitionLambda = new EventsRuleToLambda(this, 'NightlyCreateParition', {
            existingLambdaObj: lambdaFunc,
            eventRuleProps: {
                schedule: Schedule.expression("cron(1 0 * * ? *)")
            }
        });

        props.s3Bucket.grantReadWrite(createParitionLambda.lambdaFunction.role!);
        lambdaGluePolicy.attachToRole(createParitionLambda.lambdaFunction.role!);

        const createPartitionCustomResource = new CustomResource(this, 'CustomResource', {
            serviceToken: lambdaFunc.functionArn,
            resourceType: 'Custom::CreatePartition'
        });

        createParitionLambda.lambdaFunction.node.addDependency(lambdaGluePolicy);
    }
}