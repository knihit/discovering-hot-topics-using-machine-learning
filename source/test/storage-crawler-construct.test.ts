/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { Database, DataFormat, Schema, Table } from '@aws-cdk/aws-glue';
import { Bucket } from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
import { StorageCrawler } from '../lib/storage/storage-crawler-construct';


test('test orchestration construct', () => {
    const stack = new cdk.Stack();

    const db = new Database(stack, 'testDB', {
        databaseName: 'testDB'
    });

    const testBucket = new Bucket(stack, 'testBucket');

    const storageCofig: Map<string, string> = new Map();

        storageCofig.set('Sentiment', new Table(stack, 'sentiment', {
            tableName: 'sentiment',
            database: db,
            bucket: testBucket,
            s3Prefix: 'sentiment/',
            columns: [{
                name: 'field1',
                type: Schema.STRING
            }, {
                name: 'field2',
                type: Schema.STRING
            }],
            dataFormat: DataFormat.PARQUET,
            storedAsSubDirectories: true,
            partitionKeys: [{
                name: 'created_at',
                type: Schema.TIMESTAMP
            }]
        }).tableName);
        storageCofig.set('Entity', new Table(stack, 'entity', {
            tableName: 'entity',
            database: db,
            bucket: testBucket,
            s3Prefix: 'entity/',
            columns: [{
                name: 'field1',
                type: Schema.STRING
            }, {
                name: 'field2',
                type: Schema.STRING
            }],
            dataFormat: DataFormat.PARQUET,
            storedAsSubDirectories: true,
            partitionKeys: [{
                name: 'created_at',
                type: Schema.TIMESTAMP
            }]
        }).tableName);

    new StorageCrawler (stack, 'DatabaseCrawler', {
        s3Bucket: testBucket,
        databaseName: db.databaseName,
        keyArn: 'arn:aws:kms:fakeregion:fakeaccountid:kms/testkey',
        tableMap: storageCofig
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});