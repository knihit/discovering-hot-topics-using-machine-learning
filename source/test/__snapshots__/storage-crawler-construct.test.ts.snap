// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test orchestration construct 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters53b3d6bd9b1a6c0426d2d61d72b98e21f39d8fd769ba87b10b62e1ee17910529ArtifactHash27C67F5C": Object {
      "Description": "Artifact hash for asset \\"53b3d6bd9b1a6c0426d2d61d72b98e21f39d8fd769ba87b10b62e1ee17910529\\"",
      "Type": "String",
    },
    "AssetParameters53b3d6bd9b1a6c0426d2d61d72b98e21f39d8fd769ba87b10b62e1ee17910529S3BucketC579C769": Object {
      "Description": "S3 bucket for asset \\"53b3d6bd9b1a6c0426d2d61d72b98e21f39d8fd769ba87b10b62e1ee17910529\\"",
      "Type": "String",
    },
    "AssetParameters53b3d6bd9b1a6c0426d2d61d72b98e21f39d8fd769ba87b10b62e1ee17910529S3VersionKey7C229637": Object {
      "Description": "S3 key for asset version \\"53b3d6bd9b1a6c0426d2d61d72b98e21f39d8fd769ba87b10b62e1ee17910529\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "DatabaseCrawlerCustomResource3816EFB4": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "DatabaseCrawlerLambdaFunction62DC17B4",
            "Arn",
          ],
        },
      },
      "Type": "Custom::CreatePartition",
      "UpdateReplacePolicy": "Delete",
    },
    "DatabaseCrawlerLambdaFunction62DC17B4": Object {
      "DependsOn": Array [
        "DatabaseCrawlerLambdaFunctionServiceRoleDefaultPolicy2BA171D3",
        "DatabaseCrawlerLambdaFunctionServiceRole508963D0",
        "DatabaseCrawlerLambdaGlueBD6E669D",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "Lambda functions has the required permission to write CloudWatch Logs. It uses custom policy instead of arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole with more tighter permissions.",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters53b3d6bd9b1a6c0426d2d61d72b98e21f39d8fd769ba87b10b62e1ee17910529S3BucketC579C769",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters53b3d6bd9b1a6c0426d2d61d72b98e21f39d8fd769ba87b10b62e1ee17910529S3VersionKey7C229637",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters53b3d6bd9b1a6c0426d2d61d72b98e21f39d8fd769ba87b10b62e1ee17910529S3VersionKey7C229637",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "Lambda function to create parition on glue tables",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "DATABASE_NAME": Object {
              "Ref": "testDB74E9B281",
            },
            "TABLE_NAMES": Object {
              "Fn::Join": Array [
                "",
                Array [
                  Object {
                    "Ref": "sentimentTable95617508",
                  },
                  ",",
                  Object {
                    "Ref": "entityTableF0612B70",
                  },
                ],
              ],
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "DatabaseCrawlerLambdaFunctionServiceRole508963D0",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 300,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "DatabaseCrawlerLambdaFunctionLambdaInvokePermission619FECBF": Object {
      "DependsOn": Array [
        "DatabaseCrawlerLambdaGlueBD6E669D",
      ],
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "DatabaseCrawlerLambdaFunction62DC17B4",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "DatabaseCrawlerNightlyCreateParitionEventsRule82FD2B6A",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "DatabaseCrawlerLambdaFunctionServiceRole508963D0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaFunctionServiceRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DatabaseCrawlerLambdaFunctionServiceRoleDefaultPolicy2BA171D3": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "Lambda needs the following minimum required permissions to send trace data to X-Ray.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testBucketDF4D7D1A",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testBucketDF4D7D1A",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DatabaseCrawlerLambdaFunctionServiceRoleDefaultPolicy2BA171D3",
        "Roles": Array [
          Object {
            "Ref": "DatabaseCrawlerLambdaFunctionServiceRole508963D0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DatabaseCrawlerLambdaGlueBD6E669D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "glue:GetPartition",
                "glue:GetTable",
                "glue:CreatePartition",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":glue:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      Object {
                        "Ref": "testDB74E9B281",
                      },
                      "/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":glue:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":database/",
                      Object {
                        "Ref": "testDB74E9B281",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":glue:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":catalog",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Resource": "arn:aws:kms:fakeregion:fakeaccountid:kms/testkey",
            },
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":/aws-glue/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DatabaseCrawlerLambdaGlueBD6E669D",
        "Roles": Array [
          Object {
            "Ref": "DatabaseCrawlerLambdaFunctionServiceRole508963D0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DatabaseCrawlerNightlyCreateParitionEventsRule82FD2B6A": Object {
      "Properties": Object {
        "ScheduleExpression": "cron(1 0 * * ? *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "DatabaseCrawlerLambdaFunction62DC17B4",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "entityTableF0612B70": Object {
      "Properties": Object {
        "CatalogId": Object {
          "Ref": "AWS::AccountId",
        },
        "DatabaseName": Object {
          "Ref": "testDB74E9B281",
        },
        "TableInput": Object {
          "Description": "entity generated by CDK",
          "Name": "entity",
          "Parameters": Object {
            "classification": "parquet",
            "has_encrypted_data": false,
          },
          "PartitionKeys": Array [
            Object {
              "Name": "created_at",
              "Type": "timestamp",
            },
          ],
          "StorageDescriptor": Object {
            "Columns": Array [
              Object {
                "Name": "field1",
                "Type": "string",
              },
              Object {
                "Name": "field2",
                "Type": "string",
              },
            ],
            "Compressed": false,
            "InputFormat": "org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat",
            "Location": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "s3://",
                  Object {
                    "Ref": "testBucketDF4D7D1A",
                  },
                  "/entity/",
                ],
              ],
            },
            "OutputFormat": "org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat",
            "SerdeInfo": Object {
              "SerializationLibrary": "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe",
            },
            "StoredAsSubDirectories": true,
          },
          "TableType": "EXTERNAL_TABLE",
        },
      },
      "Type": "AWS::Glue::Table",
    },
    "sentimentTable95617508": Object {
      "Properties": Object {
        "CatalogId": Object {
          "Ref": "AWS::AccountId",
        },
        "DatabaseName": Object {
          "Ref": "testDB74E9B281",
        },
        "TableInput": Object {
          "Description": "sentiment generated by CDK",
          "Name": "sentiment",
          "Parameters": Object {
            "classification": "parquet",
            "has_encrypted_data": false,
          },
          "PartitionKeys": Array [
            Object {
              "Name": "created_at",
              "Type": "timestamp",
            },
          ],
          "StorageDescriptor": Object {
            "Columns": Array [
              Object {
                "Name": "field1",
                "Type": "string",
              },
              Object {
                "Name": "field2",
                "Type": "string",
              },
            ],
            "Compressed": false,
            "InputFormat": "org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat",
            "Location": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "s3://",
                  Object {
                    "Ref": "testBucketDF4D7D1A",
                  },
                  "/sentiment/",
                ],
              ],
            },
            "OutputFormat": "org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat",
            "SerdeInfo": Object {
              "SerializationLibrary": "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe",
            },
            "StoredAsSubDirectories": true,
          },
          "TableType": "EXTERNAL_TABLE",
        },
      },
      "Type": "AWS::Glue::Table",
    },
    "testBucketDF4D7D1A": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testDB74E9B281": Object {
      "Properties": Object {
        "CatalogId": Object {
          "Ref": "AWS::AccountId",
        },
        "DatabaseInput": Object {
          "Name": "testDB",
        },
      },
      "Type": "AWS::Glue::Database",
    },
  },
}
`;
